data-prepper-plugins/anomaly-detector-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/anomalydetector/AnomalyDetectorProcessor.java

```diff
@@ -0,0 +1,140 @@
+        int shingleSize = anomalyDetectorProcessorConfig.getShingleSize();
+        double timeDecay = anomalyDetectorProcessorConfig.getTimeDecay();
+	    Precision precision = Precision.FLOAT_32;
+	    int dataSize = 4 * sampleSize;
+	    int dimensions = baseDimensions * shingleSize;
+	    TransformMethod transformMethod = TransformMethod.NORMALIZE;
+
+	    forest = ThresholdedRandomCutForest.builder()
+                    .compact(true)
+                    .dimensions(dimensions)
+			        .randomSeed(0)
+                    .numberOfTrees(NUMBER_OF_TREES)
+                    .shingleSize(shingleSize)
+                    .sampleSize(sampleSize)
+			        .internalShinglingEnabled(true)
+                    .precision(precision)
+                    .anomalyRate(ANOMALY_RATE)
+                    .forestMode(ForestMode.STANDARD)
+			        .transformMethod(transformMethod).outputAfter(TRANSFORM_OUTPUT_AFTER)
+                    .timeDecay(timeDecay / sampleSize)
+			        .initialAcceptFraction(INITIAL_ACCEPT_FRACTION).build();
+	    forest.setLowerThreshold(LOWER_THRESHOLD);
+	    forest.setHorizon(HORIZON_VALUE);
+    }
+
+    @Override
+    public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
+        List<Record<Event>> recordsOut = new ArrayList<>();
+        int timeStamp = (int)Instant.now().getEpochSecond();
+        for(final Record<Event> record : records) {
+	        Event event = record.getData();
+            boolean notFound = false;
+            double[] points = new double[keys.size()];
+            int index = 0;
+            for (final String key: keys) {
+                Number value = event.get(key, Number.class);
+                if (value == null) {
+                    notFound = true;
+                    break;
+                }
+                if (value instanceof Long) {
+                    points[index] = (double)value.longValue();
+                } else if (value instanceof Integer) {
+                    points[index] = (double)value.intValue();
+                } else if (value instanceof Short) {
+                    points[index] = (double)value.shortValue();
+                } else if (value instanceof Byte) {
+                    points[index] = (double)value.byteValue();
+                } else if (value instanceof Float) {
+                    points[index] = (double)value.floatValue();
+                } else { // double
+                    points[index] = (double)value.doubleValue();
+                }
+                index++;
+            }
+            if (notFound) {
+                continue;
+            }
+	        try {
+                AnomalyDescriptor result = forest.process(points, timeStamp);
+                if ((result.getAnomalyGrade() != 0) && (result.isExpectedValuesPresent())) {
+                    if (result.getRelativeIndex() != 0 && result.isStartOfAnomaly()) {
+                        event.put(DEVIATION_KEY, result.getPastValues()[0]);
```

Comment:
This is for 1D data (when baseDimension =1, as was in the example) -- in general the issue can be any of the indices in 80-104. But that makes DEVIATION_KEY as type float[]. let me double check if pastValues is always non-Null (it has been a while).